streamingccl: span use-after-finish

**Describe the problem**

Seen during a recent stress test:

```
panic: use of Span after Finish. Span: stream-ingestion-flush. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 36753 [running]:
panic({0x5c09ca0, 0xc0012e2570})
        GOROOT/src/runtime/panic.go:987 +0x3ba fp=0xc006818ab0 sp=0xc0068189f0 pc=0x498a3a
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc15f6b8?)
        github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x12e fp=0xc006818b20 sp=0xc006818ab0 pc=0xcaa96e
github.com/cockroachdb/cockroach/pkg/util/tracing.WithParent(0xc003a5b400)
        github.com/cockroachdb/cockroach/pkg/util/tracing/span_options.go:212 +0x27 fp=0xc006818b38 sp=0xc006818b20 pc=0xcb1027
github.com/cockroachdb/cockroach/pkg/util/tracing.EnsureChildSpan({0x796d808, 0xc006218d00}, 0xc006218d00?, {0x6232af7, 0x10}, {0x0, 0x0, 0xc006818c38?})
        github.com/cockroachdb/cockroach/pkg/util/tracing/tracer.go:1602 +0x7a fp=0xc006818bd8 sp=0xc006818b38 pc=0xcb841a
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc001d48a00, {0x796d808, 0xc006218d00}, 0xc0031f8000)
        github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:945 +0xf2 fp=0xc006818e58 sp=0xc006818bd8 pc=0x1c69852
github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send(0xc000f97d28, {0x796d808, 0xc006218d00}, 0xc0031f8000)
        github.com/cockroachdb/cockroach/pkg/kv/db.go:224 +0xa6 fp=0xc006818ee0 sp=0xc006818e58 pc=0x17a3106
github.com/cockroachdb/cockroach/pkg/kv/bulk.(*SSTBatcher).addSSTable.func1({0x796d808, 0xc006218d00}, 0xc009cdf4a0, 0xc002829400, {0x7f0316ab2688?, 0x12d22a0?, 0xc0?}, 0xc008ce8dd0, 0x0, 0xc0068197e0, ...)
        github.com/cockroachdb/cockroach/pkg/kv/bulk/sst_batcher.go:820 +0x630 fp=0xc006819700 sp=0xc006818ee0 pc=0x3c45130
github.com/cockroachdb/cockroach/pkg/kv/bulk.(*SSTBatcher).addSSTable(0xc002266848?, {0x796d808, _}, {_, _, _}, {_, _, _}, {0xc00ef53f00, ...}, ...)
        github.com/cockroachdb/cockroach/pkg/kv/bulk/sst_batcher.go:904 +0x913 fp=0xc006819c60 sp=0xc006819700 pc=0x3c44753
github.com/cockroachdb/cockroach/pkg/kv/bulk.(*SSTBatcher).doFlush.func1({0x796d808?, 0xc006218d00?})
        github.com/cockroachdb/cockroach/pkg/kv/bulk/sst_batcher.go:657 +0x2f3 fp=0xc006819f58 sp=0xc006819c60 pc=0x3c43473
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25 fp=0xc006819f78 sp=0xc006819f58 pc=0x1c56a85
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64 fp=0xc006819fe0 sp=0xc006819f78 pc=0x1b9f844
runtime.goexit()
        GOROOT/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc006819fe8 sp=0xc006819fe0 pc=0x4cdbc1
created by golang.org/x/sync/errgroup.(*Group).Go
        golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
```


Jira issue: CRDB-31264

--- a/pkg/ccl/backupccl/restore_data_processor.go
+++ b/pkg/ccl/backupccl/restore_data_processor.go
@@ -767,7 +767,7 @@ func reserveRestoreWorkerMemory(
 // implement a mock SSTBatcher used purely for job progress tracking.
 type SSTBatcherExecutor interface {
 	AddMVCCKey(ctx context.Context, key storage.MVCCKey, value []byte) error
-	Reset(ctx context.Context) error
+	Reset(ctx context.Context)
 	Flush(ctx context.Context) error
 	Close(ctx context.Context)
 	GetSummary() kvpb.BulkOpSummary
@@ -786,9 +786,7 @@ func (b *sstBatcherNoop) AddMVCCKey(ctx context.Context, key storage.MVCCKey, va
 }
 
 // Reset resets the counter
-func (b *sstBatcherNoop) Reset(ctx context.Context) error {
-	return nil
-}
+func (b *sstBatcherNoop) Reset(ctx context.Context) {}
 
 // Flush noops.
 func (b *sstBatcherNoop) Flush(ctx context.Context) error {

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go
@@ -1188,9 +1188,7 @@ func (sip *streamIngestionProcessor) flushBuffer(b flushableBuffer) (*jobspb.Res
 	sip.metrics.IngestedEvents.Inc(int64(len(b.buffer.curKVBatch)))
 	sip.metrics.IngestedEvents.Inc(int64(len(b.buffer.curRangeKVBatch)))
 
-	if err := sip.batcher.Reset(ctx); err != nil {
-		return b.checkpoint, err
-	}
+	sip.batcher.Reset(ctx)
 
 	releaseBuffer(b.buffer)
 

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go
@@ -1155,6 +1155,8 @@ type flushableBuffer struct {
 func (sip *streamIngestionProcessor) flushBuffer(b flushableBuffer) (*jobspb.ResolvedSpans, error) {
 	ctx, sp := tracing.ChildSpan(sip.Ctx(), "stream-ingestion-flush")
 	defer sp.Finish()
+	// Ensure the batcher is always reset, even on early error returns.
+	defer sip.batcher.Reset(ctx)
 
 	// First process the point KVs.
 	//
@@ -1187,8 +1189,6 @@ func (sip *streamIngestionProcessor) flushBuffer(b flushableBuffer) (*jobspb.Res
 	sip.metrics.Flushes.Inc(1)
 	sip.metrics.IngestedEvents.Inc(int64(len(b.buffer.curKVBatch)))
 	sip.metrics.IngestedEvents.Inc(int64(len(b.buffer.curRangeKVBatch)))
-
-	sip.batcher.Reset(ctx)
 
 	releaseBuffer(b.buffer)
 

--- a/pkg/kv/bulk/buffering_adder.go
+++ b/pkg/kv/bulk/buffering_adder.go
@@ -256,9 +256,7 @@ func (b *BufferingAdder) doFlush(ctx context.Context, forSize bool) error {
 		b.curBufSummary.Reset()
 		return nil
 	}
-	if err := b.sink.Reset(ctx); err != nil {
-		return err
-	}
+	b.sink.Reset(ctx)
 	b.sink.currentStats.BufferFlushes++
 
 	var before *bulkpb.IngestionPerformanceStats

--- a/pkg/kv/bulk/sst_batcher.go
+++ b/pkg/kv/bulk/sst_batcher.go
@@ -374,7 +374,13 @@ func (b *SSTBatcher) AddMVCCKey(ctx context.Context, key storage.MVCCKey, value 
 
 // Reset clears all state in the batcher and prepares it for reuse.
 func (b *SSTBatcher) Reset(ctx context.Context) {
+	if err := b.asyncAddSSTs.Wait(); err != nil {
+		log.Warningf(ctx, "closing with flushes in-progress encountered an error: %v", err)
+	}
+	b.asyncAddSSTs = ctxgroup.Group{}
+
 	b.sstWriter.Close()
+
 	b.sstFile = &storage.MemObject{}
 	// Create sstables intended for ingestion using the newest format that all
 	// nodes can support. MakeIngestionSSTWriter will handle cluster version

--- a/pkg/kv/bulk/sst_batcher.go
+++ b/pkg/kv/bulk/sst_batcher.go
@@ -232,8 +232,8 @@ func MakeSSTBatcher(
 	}
 	b.mu.lastFlush = timeutil.Now()
 	b.mu.tracingSpan = tracing.SpanFromContext(ctx)
-	err := b.Reset(ctx)
-	return b, err
+	b.Reset(ctx)
+	return b, nil
 }
 
 // MakeStreamSSTBatcher creates a batcher configured to ingest duplicate keys
@@ -258,8 +258,8 @@ func MakeStreamSSTBatcher(
 	b.mu.lastFlush = timeutil.Now()
 	b.mu.tracingSpan = tracing.SpanFromContext(ctx)
 	b.SetOnFlush(onFlush)
-	err := b.Reset(ctx)
-	return b, err
+	b.Reset(ctx)
+	return b, nil
 }
 
 // MakeTestingSSTBatcher creates a batcher for testing, allowing setting options
@@ -281,8 +281,8 @@ func MakeTestingSSTBatcher(
 		mem:            mem,
 		limiter:        sendLimiter,
 	}
-	err := b.Reset(ctx)
-	return b, err
+	b.Reset(ctx)
+	return b, nil
 }
 
 func (b *SSTBatcher) updateMVCCStats(key storage.MVCCKey, value []byte) {
@@ -373,7 +373,7 @@ func (b *SSTBatcher) AddMVCCKey(ctx context.Context, key storage.MVCCKey, value 
 }
 
 // Reset clears all state in the batcher and prepares it for reuse.
-func (b *SSTBatcher) Reset(ctx context.Context) error {
+func (b *SSTBatcher) Reset(ctx context.Context) {
 	b.sstWriter.Close()
 	b.sstFile = &storage.MemObject{}
 	// Create sstables intended for ingestion using the newest format that all
@@ -402,8 +402,6 @@ func (b *SSTBatcher) Reset(ctx context.Context) error {
 	if b.mu.totalStats.SendWaitByStore == nil {
 		b.mu.totalStats.SendWaitByStore = make(map[roachpb.StoreID]time.Duration)
 	}
-
-	return nil
 }
 
 const (
@@ -438,7 +436,8 @@ func (b *SSTBatcher) flushIfNeeded(ctx context.Context, nextKey roachpb.Key) err
 		if err := b.doFlush(ctx, rangeFlush); err != nil {
 			return err
 		}
-		return b.Reset(ctx)
+		b.Reset(ctx)
+		return nil
 	}
 
 	if b.sstWriter.DataSize >= ingestFileSize(b.settings) {
@@ -462,7 +461,8 @@ func (b *SSTBatcher) flushIfNeeded(ctx context.Context, nextKey roachpb.Key) err
 		if err := b.doFlush(ctx, sizeFlush); err != nil {
 			return err
 		}
-		return b.Reset(ctx)
+		b.Reset(ctx)
+		return nil
 	}
 	return nil
 }

